sql/core/src/main/scala/org/apache/spark/sql/execution/commands/parsers.scala

```diff
@@ -0,0 +1,420 @@
+      val allowExisting = base.getClauseOption("TOK_IFEXISTS", args)
+      AlterTableDropProperties(
+        tableIdent,
+        dropTableProperties, allowExisting.isDefined)(node.source)
+
+    case Token("TOK_ALTERTABLE_SERIALIZER", serdeArgs) :: rest =>
+      val serdeClassName = unquoteString(cleanIdentifier(serdeArgs.head.text))
+
+      val serdeProperties: Option[Map[String, Option[String]]] = Option(
+        // SET SERDE serde_classname WITH SERDEPROPERTIES
+        if (serdeArgs.tail.isEmpty) {
+          null
+        } else {
+          extractTableProps(serdeArgs.tail.head)
+        }
+      )
+
+      AlterTableSerDeProperties(
+        tableIdent,
+        Some(serdeClassName),
+        serdeProperties,
+        partition)(node.source)
+
+    case Token("TOK_ALTERTABLE_SERDEPROPERTIES", args) :: rest =>
+      val serdeProperties: Map[String, Option[String]] = extractTableProps(args.head)
+
+      AlterTableSerDeProperties(
+        tableIdent,
+        None,
+        Some(serdeProperties),
+        partition)(node.source)
+
+    case (bucketSpec @ Token("TOK_ALTERTABLE_CLUSTER_SORT", _)) :: rest =>
+      val (buckets, noClustered, noSorted) = bucketSpec match {
+        case Token("TOK_ALTERTABLE_CLUSTER_SORT", clusterAndSoryByArgs :: Nil) =>
+          clusterAndSoryByArgs match {
+              case Token("TOK_ALTERTABLE_BUCKETS", bucketArgs) =>
+                val bucketCols = bucketArgs.head.children.map(_.text)
+
+                val (sortCols, sortDirections, numBuckets) = {
+                  if (bucketArgs(1).text == "TOK_TABCOLNAME") {
+                    val cols = bucketArgs(1).children.map {
+                      case Token("TOK_TABSORTCOLNAMEASC", Token(colName, Nil) :: Nil) =>
+                        (colName, Ascending)
+                      case Token("TOK_TABSORTCOLNAMEDESC", Token(colName, Nil) :: Nil) =>
+                        (colName, Descending)
+                    }
+                    (cols.map(_._1), cols.map(_._2), bucketArgs(2).text.toInt)
+                  } else {
+                    (Nil, Nil, bucketArgs(1).text.toInt)
+                  }
+                }
+
+                (Some(BucketSpec(numBuckets, bucketCols, sortCols, sortDirections)),
+                  false, false)
+              case Token("TOK_NOT_CLUSTERED", Nil) =>
+                (None, true, false)
+              case Token("TOK_NOT_SORTED", Nil) =>
+                (None, false, true)
+          }
+      }
+
+      AlterTableStoreProperties(
+        tableIdent,
+        buckets,
+        noClustered,
+        noSorted)(node.source)
+
+    case Token("TOK_ALTERTABLE_BUCKETS", Token(bucketNum, Nil) :: Nil) :: rest =>
+      val num = bucketNum.toInt
+      val buckets = Some(BucketSpec(num, Nil, Nil, Nil))
+      AlterTableStoreProperties(
+        tableIdent,
+        buckets,
+        false,
+        false)(node.source)
+
+    case (tableSkewed @ Token("TOK_ALTERTABLE_SKEWED", _)) :: rest =>
```

Comment:
Maybe it is easier to separate the skewed/not-skewed cases here... Why not extract the children, and do a match on those?